File: java/src/main/java/dev/csvdiff/Profiles.java
package dev.csvdiff;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import org.tomlj.Toml;
import org.tomlj.TomlArray;
import org.tomlj.TomlParseResult;
import org.tomlj.TomlTable;

/**
 * Named comparison profiles from {@code csvdiff.toml}, so a recurring comparison does not get
 * retyped. Searched in the working directory, then {@code ~/.config/csvdiff/}. The file format is
 * shared with the Python and TypeScript implementations:
 *
 * <pre>{@code
 * [profiles.orders]
 * key       = ["order_id", "line_no"]
 * compare   = ["qty", "price", "status"]   # omit for all common non-key columns
 * ignore    = ["updated_at"]
 * trim      = true
 * tolerance = 0.005
 * }</pre>
 */
public final class Profiles {

  private Profiles() {}

  /** Where a config file is looked for when no path is given. */
  public static List<Path> searchPath() {
    return List.of(
        Path.of("csvdiff.toml"),
        Path.of(System.getProperty("user.home"), ".config", "csvdiff", "csvdiff.toml"));
  }

  /** Loads the config, or an empty one when no file exists. */
  public static TomlParseResult load(String explicitPath) {
    List<Path> candidates = explicitPath != null ? List.of(Path.of(explicitPath)) : searchPath();
    for (Path p : candidates) {
      if (Files.isRegularFile(p)) {
        try {
          TomlParseResult toml = Toml.parse(p);
          if (toml.hasErrors()) {
            String first = toml.errors().getFirst().toString();
            throw new CsvDiffException("Cannot parse " + p + ": " + first);
          }
          return toml;
        } catch (IOException e) {
          throw new UncheckedIOException("Cannot read " + p, e);
        }
      }
    }
    return Toml.parse("");
  }

  /**
   * Looks up one profile.
   *
   * @throws CsvDiffException if the name was given but no such profile exists
   */
  public static Optional<TomlTable> profile(TomlParseResult config, String name) {
    if (name == null || name.isBlank()) {
      return Optional.empty();
    }
    TomlTable profiles = config.getTable("profiles");
    TomlTable found = profiles == null ? null : profiles.getTable(name);
    if (found == null) {
      throw new CsvDiffException("Profile not found: " + name);
    }
    return Optional.of(found);
  }

  /** Applies a profile's values as the base layer of an {@link Options.Builder}. */
  public static Options.Builder apply(Options.Builder builder, TomlTable p) {
    return builder
        .key(strings(p, "key"))
        .compare(strings(p, "compare"))
        .ignore(strings(p, "ignore"))
        .trim(p.getBoolean("trim"))
        .ignoreCase(p.getBoolean("ignore_case"))
        .emptyIsNull(p.getBoolean("empty_is_null"))
        .tolerance(p.getDouble("tolerance"))
        .maxRows(intOf(p, "max_rows"))
        .delimiter(charOf(p, "delimiter"))
        .encoding(p.getString("encoding"))
        .engine(p.getString("engine"));
  }

  private static List<String> strings(TomlTable t, String key) {
    TomlArray arr = t.getArray(key);
    if (arr == null) {
      return null;
    }
    var out = new ArrayList<String>(arr.size());
    for (int i = 0; i < arr.size(); i++) {
      out.add(String.valueOf(arr.get(i)));
    }
    return List.copyOf(out);
  }

  private static Integer intOf(TomlTable t, String key) {
    Long v = t.getLong(key);
    return v == null ? null : Math.toIntExact(v);
  }

  private static Character charOf(TomlTable t, String key) {
    String v = t.getString(key);
    return v == null || v.isEmpty() ? null : v.charAt(0);
  }

  /** Splits {@code "a, b ,c"} into a list; returns {@code null} for {@code null}. */
  public static List<String> parseList(String value) {
    if (value == null) {
      return null;
    }
    return java.util.Arrays.stream(value.split(","))
        .map(String::strip)
        .filter(s -> !s.isEmpty())
        .toList();
  }
}

File: rust/tests/profiles.rs
//! csvdiff.toml profiles, which the other implementations read too.

use std::fs;

use csvdiff::options::Options;
use csvdiff::profiles;

const CONFIG: &str = r#"
# a comment
[profiles.orders]
key       = ["order_id", "line_no"]
compare   = ["qty", "price"]
ignore    = ["updated_at"]   # trailing comment
trim      = true
tolerance = 0.005
max_rows  = 100
engine    = "native"

[profiles.other]
key = ["id"]
"#;

#[test]
fn a_profile_becomes_the_base_layer_of_the_options() {
    let found = profiles::parse(CONFIG).expect("parse");
    assert_eq!(found.len(), 2);

    let mut opt = Options::default();
    found["orders"].apply_to(&mut opt);
    assert_eq!(opt.key, vec!["order_id", "line_no"]);
    assert_eq!(
        opt.compare,
        Some(vec!["qty".to_string(), "price".to_string()])
    );
    assert_eq!(
        opt.ignore,
        vec!["updated_at"],
        "a trailing comment leaked into ignore"
    );
    assert!(opt.trim);
    assert_eq!(opt.tolerance, 0.005);
    assert_eq!(opt.max_rows, 100);
    assert_eq!(opt.engine, "native");

    // Anything given on the command line is applied after, and wins.
    opt.max_rows = 7;
    assert_eq!(opt.max_rows, 7);
}

#[test]
fn an_explicit_config_path_that_does_not_exist_is_an_error() {
    let missing = std::env::temp_dir().join("csvdiff-no-such-config.toml");
    let _ = fs::remove_file(&missing);
    assert!(profiles::load(Some(&missing.to_string_lossy())).is_err());
}

#[test]
fn parse_list_drops_empty_entries() {
    assert_eq!(csvdiff::parse_list(" a, b ,,c "), vec!["a", "b", "c"]);
    assert!(csvdiff::parse_list("  ").is_empty());
}
